compiler/rustc_lint/src/useless_send_constraint.rs

```diff
@@ -0,0 +1,51 @@
+declare_lint! {
+    /// The `lint_useless_send_constraint` lints useless constraint of references to `Send`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// fn foo(_: &(dyn Any + Send>) {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// References cannot be sent across threads, so constraining them to `Send` is useless.
```

Comment:
This statement is wrong. If the base type implements [`Sync`](https://doc.rust-lang.org/stable/std/marker/trait.Sync.html) then the reference *can* be sent across threads.

The issue is that references of `dyn (Send + ...)` can't be sent across threads, at least as long as `...` doesn't contain a `Sync` bound.

Reply:
Thanks have addressed

Reply:
Just to clarify the behaviour:

* `&dyn Send` should be changed to `&dyn Any`
* `&dyn ... + Send` should have send removed, unless sync is specified
* Mutable references should be ignored

Have only been learning rust for a couple of weeks so my understanding of the threading model is a bit weak, but I will read more about it to try and improve my understanding

Reply:
> Have only been learning rust for a couple of weeks

Don't worry, we are all learners!

The `Send` bound is never useful. If `Sync` is not specified, adding `Send` won't make the reference `Send`, it will always not be `Send`. If `Sync` is specified, `Send` is always implemented whether it's specified or not.

I'm not sure actually what's going on with mutable references, what would be your reasoning why they should be ignored?

Reply:
> The Send bound is never useful. If Sync is not specified, adding Send won't make the reference Send, it will always not be Send. If Sync is specified, Send is always implemented whether it's specified or not.

Ah right that makes sense, like with copy and clone. When it's `Sync + Send`, I'll give a different message that it's redundant due to sync being a super trait

> I'm not sure actually what's going on with mutable references, what would be your reasoning why they should be ignored?

I don't think there was any real logic behind it, I think I assumed based off of the original issue + the fact it seemed to break a lot of code when I removed a `+ Send` on a `*mut (dyn Any + Send)`, whereas when I removed it on non-mut references it didn't break anything, but that must've just been a coincidence...

Reply:
> I'm not sure actually what's going on with mutable references

`&mut T` is `Send` iff `T` is `Send`, even though this is kind of badly documented. [(proof)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=190547f8744598ca50fd19ba8aa3654c)

So this lint should only trigger for exactly `&dyn Send`, not `&mut dyn Send`.

Reply:
To clarify: in `&dyn X + Send`, neither the `X` nor the `Send` is about the reference. It means that the thing that the reference points at implements both `X` and `Send`. But whether or not something implements `Send` isn't relevant if you only have a reference to it, because you can't send it if you don't own it.

Reply:
I've made a PR to add it to the docs: #111065